File: CBLClient/Apps/JavaRequestHandlers/src/main/java/com/couchbase/lite/mobiletestkit/legacy/ReplicatorConfigurationRequestHandler.java
package com.couchbase.lite.mobiletestkit.legacy;


import java.io.IOException;
import java.io.InputStream;
import java.net.URI;
import java.net.URISyntaxException;
import java.util.EnumSet;
import java.util.List;
import java.util.Map;

import com.couchbase.lite.Authenticator;
import com.couchbase.lite.Conflict;
import com.couchbase.lite.ConflictResolver;
import com.couchbase.lite.Database;
import com.couchbase.lite.DatabaseEndpoint;
import com.couchbase.lite.Document;
import com.couchbase.lite.DocumentFlag;
import com.couchbase.lite.MutableDocument;
import com.couchbase.lite.ReplicationFilter;
import com.couchbase.lite.ReplicatorConfiguration;
import com.couchbase.lite.ReplicatorType;
import com.couchbase.lite.URLEndpoint;
import com.couchbase.lite.mobiletestkit.Args;
import com.couchbase.lite.mobiletestkit.TestKitApp;
import com.couchbase.lite.mobiletestkit.util.FileUtils;
import com.couchbase.lite.mobiletestkit.util.Log;

import static java.lang.Thread.sleep;


public class ReplicatorConfigurationRequestHandler {
    private static final String TAG = "REPLCONFIGHANDLER";

    public ReplicatorConfiguration builderCreate(Args args) throws URISyntaxException {
        Database sourceDb = args.get("sourceDb", Database.class);
        Database targetDb = args.get("targetDb", Database.class);
        URI targetURI = null;
        if (args.getString("targetURI") != null) {
            targetURI = new URI(args.getString("targetURI"));
        }
        if (targetDb != null) {
            DatabaseEndpoint target = new DatabaseEndpoint(targetDb);
            return new ReplicatorConfiguration(sourceDb, target);
        }
        else if (targetURI != null) {
            URLEndpoint target = new URLEndpoint(targetURI);
            return new ReplicatorConfiguration(sourceDb, target);
        }
        else {
            throw new IllegalArgumentException("Incorrect configuration parameter provided");
        }
    }

    public ReplicatorConfiguration configure(Args args) throws URISyntaxException, IOException {
        Database sourceDb = args.get("source_db", Database.class);
        String url = args.getString("target_url");
        URI targetURL = (url == null) ? null : new URI(args.getString("target_url"));
        Database targetDb = args.get("target_db", Database.class);
        String replicatorType = args.getString("replication_type");
        Boolean continuous = args.getBoolean("continuous");
        List<String> channels = args.getList("channels");
        List<String> documentIds = args.getList("documentIDs");
        String pinnedservercert = args.getString("pinnedservercert");
        Authenticator authenticator = args.get("authenticator", Authenticator.class);
        Boolean push_filter = args.getBoolean("push_filter");
        Boolean pull_filter = args.getBoolean("pull_filter");
        String filter_callback_func = args.getString("filter_callback_func");
        String conflict_resolver = args.getString("conflict_resolver");
        Map<String, String> headers = args.getMap("headers");
        String heartbeat = args.getString("heartbeat");
        String maxRetries = args.getString("max_retries");
        String maxRetryWaitTime = args.getString("max_timeout");
        String auto_purge = args.getString("auto_purge");

        if (replicatorType == null) {
            replicatorType = "push_pull";
        }
        replicatorType = replicatorType.toLowerCase();
        ReplicatorType replType;
        if (replicatorType.equals("push")) {
            replType = ReplicatorType.PUSH;
        }
        else if (replicatorType.equals("pull")) {
            replType = ReplicatorType.PULL;
        }
        else {
            replType = ReplicatorType.PUSH_AND_PULL;
        }
        ReplicatorConfiguration config;
        if (sourceDb != null && targetURL != null) {
            URLEndpoint target = new URLEndpoint(targetURL);
            config = new ReplicatorConfiguration(sourceDb, target);
        }
        else if (sourceDb != null && targetDb != null) {
            DatabaseEndpoint target = new DatabaseEndpoint(targetDb);
            config = new ReplicatorConfiguration(sourceDb, target);
        }
        else {
            throw new IllegalArgumentException("\"No source db provided or target url provided\"");
        }
        if (continuous != null) {
            config.setContinuous(continuous);
        }
        else {
            config.setContinuous(false);
        }
        if (headers != null) {
            config.setHeaders(headers);
        }
        if (authenticator != null) {
            config.setAuthenticator(authenticator);
        }
        config.setType(replType);
        /*if (conflictResolver != null) {
            config.setConflictResolver(conflictResolver);
        }*/
        if (channels != null) {
            config.setChannels(channels);
        }
        if (documentIds != null) {
            config.setDocumentIDs(documentIds);
        }
        if (heartbeat != null && !heartbeat.trim().isEmpty()) {
            config.setHeartbeat(Integer.parseInt(heartbeat));
        }
        if (maxRetries != null && !maxRetries.trim().isEmpty()) {
            config.setMaxAttempts(Integer.parseInt(maxRetries));
        }
        if (maxRetryWaitTime != null && !maxRetryWaitTime.trim().isEmpty()) {
            config.setMaxAttemptWaitTime(Integer.parseInt(maxRetryWaitTime));
        }
        if (auto_purge != null) {
            if (auto_purge.equalsIgnoreCase("enabled")) {
                Log.i(TAG, "auto purge is enabled explicitly");
                config.setAutoPurgeEnabled(true);
            }
            else if (auto_purge.equalsIgnoreCase("disabled")) {
                Log.i(TAG, "auto purge is disabled");
                config.setAutoPurgeEnabled(false);
            }
        }
        else {
            Log.i(TAG, "auto purge not specified, use the default setting.");
        }

        Log.d(TAG, "Args: " + args);
        if (pinnedservercert != null) {
            byte[] ServerCert = this.getPinnedCertFile();
            // Set pinned certificate.
            config.setPinnedServerCertificate(ServerCert);
        }
        if (push_filter) {
            switch (filter_callback_func) {
                case "boolean":
                    config.setPushFilter(new ReplicatorBooleanFilterCallback());
                    break;
                case "deleted":
                    config.setPushFilter(new ReplicatorDeletedFilterCallback());
                    break;
                case "access_revoked":
                    config.setPushFilter(new ReplicatorAccessRevokedFilterCallback());
                    break;
                default:
                    config.setPushFilter(new DefaultReplicatorFilterCallback());
                    break;
            }
        }
        if (pull_filter) {
            switch (filter_callback_func) {
                case "boolean":
                    config.setPullFilter(new ReplicatorBooleanFilterCallback());
                    break;
                case "deleted":
                    config.setPullFilter(new ReplicatorDeletedFilterCallback());
                    break;
                case "access_revoked":
                    config.setPullFilter(new ReplicatorAccessRevokedFilterCallback());
                    break;
                default:
                    config.setPullFilter(new DefaultReplicatorFilterCallback());
                    break;
            }
        }
        switch (conflict_resolver) {
            case "local_wins":
                config.setConflictResolver(new LocalWinsCustomConflictResolver());
                break;
            case "remote_wins":
                config.setConflictResolver(new RemoteWinsCustomConflictResolver());
                break;
            case "null":
                config.setConflictResolver(new NullCustomConflictResolver());
                break;
            case "merge":
                config.setConflictResolver(new MergeCustomConflictResolver());
                break;
            case "incorrect_doc_id":
                config.setConflictResolver(new IncorrectDocIdConflictResolver());
                break;
            case "delayed_local_win":
                config.setConflictResolver(new DelayedLocalWinConflictResolver());
                break;
            case "delete_not_win":
                config.setConflictResolver(new DeleteDocConflictResolver());
                break;
            case "exception_thrown":
                config.setConflictResolver(new ExceptionThrownConflictResolver());
                break;
            default:
                config.setConflictResolver(ConflictResolver.DEFAULT);
                break;
        }
        return config;
    }

    public ReplicatorConfiguration create(Args args) {
        return args.get("variable", ReplicatorConfiguration.class);
    }

    public Authenticator getAuthenticator(Args args) {
        ReplicatorConfiguration replicatorConfiguration = args.get("variable", ReplicatorConfiguration.class);
        return replicatorConfiguration.getAuthenticator();
    }

    public List<String> getChannels(Args args) {
        ReplicatorConfiguration replicatorConfiguration = args.get("variable", ReplicatorConfiguration.class);
        return replicatorConfiguration.getChannels();
    }

    /*public ConflictResolver getConflictResolver(Args args){
        ReplicatorConfiguration replicatorConfiguration = args.get("configuration");
        return replicatorConfiguration.getConflictResolver();
    }*/

    public Database getDatabase(Args args) {
        ReplicatorConfiguration replicatorConfiguration = args.get("variable", ReplicatorConfiguration.class);
        return replicatorConfiguration.getDatabase();
    }

    public List<String> getDocumentIDs(Args args) {
        ReplicatorConfiguration replicatorConfiguration = args.get("variable", ReplicatorConfiguration.class);
        return replicatorConfiguration.getDocumentIDs();
    }

    public byte[] getPinnedServerCertificate(Args args) {
        ReplicatorConfiguration replicatorConfiguration = args.get("variable", ReplicatorConfiguration.class);
        return replicatorConfiguration.getPinnedServerCertificate();
    }

    public String getReplicatorType(Args args) {
        ReplicatorConfiguration replicatorConfiguration = args.get("variable", ReplicatorConfiguration.class);
        return replicatorConfiguration.getType().toString();
    }

    public String getTarget(Args args) {
        ReplicatorConfiguration replicatorConfiguration = args.get("variable", ReplicatorConfiguration.class);
        return replicatorConfiguration.getTarget().toString();
    }

    public Boolean isContinuous(Args args) {
        ReplicatorConfiguration replicatorConfiguration = args.get("variable", ReplicatorConfiguration.class);
        return replicatorConfiguration.isContinuous();
    }

    public void setAuthenticator(Args args) {
        ReplicatorConfiguration replicatorConfiguration = args.get("variable", ReplicatorConfiguration.class);
        Authenticator authenticator = args.get("authenticator", Authenticator.class);
        replicatorConfiguration.setAuthenticator(authenticator);
    }

    public void setChannels(Args args) {
        ReplicatorConfiguration replicatorConfiguration = args.get("variable", ReplicatorConfiguration.class);
        List<String> channels = args.getList("documentIds");
        replicatorConfiguration.setChannels(channels);
    }

    /*public void setConflictResolver(Args args){
        ReplicatorConfiguration replicatorConfiguration = args.get("configuration");
        ConflictResolver conflictResolver = args.get("conflictResolver");
        replicatorConfiguration.setConflictResolver(conflictResolver);
    }*/

    public void setContinuous(Args args) {
        ReplicatorConfiguration replicatorConfiguration = args.get("variable", ReplicatorConfiguration.class);
        Boolean continuous = args.getBoolean("continuous");
        replicatorConfiguration.setContinuous(continuous);
    }

    public void setDocumentIDs(Args args) {
        ReplicatorConfiguration replicatorConfiguration = args.get("variable", ReplicatorConfiguration.class);
        List<String> documentIds = args.getList("documentIds");
        replicatorConfiguration.setDocumentIDs(documentIds);
    }

    public void setPinnedServerCertificate(Args args) {
        ReplicatorConfiguration replicatorConfiguration = args.get("variable", ReplicatorConfiguration.class);
        byte[] cert = args.getData("cert");
        replicatorConfiguration.setPinnedServerCertificate(cert);
    }

    public void setReplicatorType(Args args) {
        ReplicatorConfiguration replicatorConfiguration = args.get("variable", ReplicatorConfiguration.class);
        String type = args.getString("replType");
        ReplicatorType replicatorType;
        switch (type) {
            case "PUSH":
                replicatorType = ReplicatorType.PUSH;
                break;
            case "PULL":
                replicatorType = ReplicatorType.PULL;
                break;
            default:
                replicatorType = ReplicatorType.PUSH_AND_PULL;
        }
        replicatorConfiguration.setType(replicatorType);
    }

    public void setAutoPurge(Args args) {
        ReplicatorConfiguration replicatorConfiguration = args.get("variable", ReplicatorConfiguration.class);
        Boolean auto_purge = args.getBoolean("auto_purge");
        replicatorConfiguration.setAutoPurgeEnabled(auto_purge);
    }

    private byte[] getPinnedCertFile() throws IOException {
        try (InputStream is = TestKitApp.getApp().getAsset("sg_cert.cer")) {
            return new FileUtils().readToByteArray(is);
        }
    }

    public String addCollection(Args args) {
        throw new UnsupportedOperationException("replicatorConfigurartionRequestHandler_addCollection");
    }
}

class ReplicatorBooleanFilterCallback implements ReplicationFilter {
    @Override
    public boolean filtered(Document document, EnumSet<DocumentFlag> flags) {
        String key = "new_field_1";
        if (document.contains(key)) {
            return document.getBoolean(key);
        }
        return true;
    }
}

class DefaultReplicatorFilterCallback implements ReplicationFilter {
    @Override
    public boolean filtered(Document document, EnumSet<DocumentFlag> flags) {
        return true;
    }
}

class ReplicatorDeletedFilterCallback implements ReplicationFilter {
    @Override
    public boolean filtered(Document document, EnumSet<DocumentFlag> flags) {
        return !(flags.contains(DocumentFlag.DELETED));
    }
}

class ReplicatorAccessRevokedFilterCallback implements ReplicationFilter {
    @Override
    public boolean filtered(Document document, EnumSet<DocumentFlag> flags) {
        return !(flags.contains(DocumentFlag.ACCESS_REMOVED));
    }
}

class LocalWinsCustomConflictResolver implements ConflictResolver {
    private static final String TAG = "CCRREPLCONFIGHANDLER";

    @Override
    public Document resolve(Conflict conflict) {
        Document localDoc = conflict.getLocalDocument();
        Document remoteDoc = conflict.getRemoteDocument();
        if (localDoc == null || remoteDoc == null) {
            throw new IllegalStateException("Either local doc or remote is/are null");
        }
        String docId = conflict.getDocumentId();
        Utility util_obj = new Utility();
        util_obj.checkMismatchDocId(localDoc, remoteDoc, docId);
        return localDoc;
    }
}

class RemoteWinsCustomConflictResolver implements ConflictResolver {
    private static final String TAG = "CCRREPLCONFIGHANDLER";

    @Override
    public Document resolve(Conflict conflict) {
        Document localDoc = conflict.getLocalDocument();
        Document remoteDoc = conflict.getRemoteDocument();
        if (localDoc == null || remoteDoc == null) {
            throw new IllegalStateException("Either local doc or remote is/are null");
        }
        String docId = conflict.getDocumentId();
        Utility util_obj = new Utility();
        util_obj.checkMismatchDocId(localDoc, remoteDoc, docId);
        return remoteDoc;
    }
}

class NullCustomConflictResolver implements ConflictResolver {
    private static final String TAG = "CCRREPLCONFIGHANDLER";

    @Override
    public Document resolve(Conflict conflict) {
        Document localDoc = conflict.getLocalDocument();
        Document remoteDoc = conflict.getRemoteDocument();
        if (localDoc == null || remoteDoc == null) {
            throw new IllegalStateException("Either local doc or remote is/are null");
        }
        String docId = conflict.getDocumentId();
        Utility util_obj = new Utility();
        util_obj.checkMismatchDocId(localDoc, remoteDoc, docId);
        return null;
    }
}

/**
 * Migrate the conflicted doc.
 * Algorithm creates a new doc with copying local doc and then adding any additional key
 * from remote doc. Conflicting keys will have value from local doc.
 */
class MergeCustomConflictResolver implements ConflictResolver {
    private static final String TAG = "CCRREPLCONFIGHANDLER";

    @Override
    public Document resolve(Conflict conflict) {
        Document localDoc = conflict.getLocalDocument();
        Document remoteDoc = conflict.getRemoteDocument();
        if (localDoc == null || remoteDoc == null) {
            throw new IllegalStateException("Either local doc or remote is/are null");
        }
        String docId = conflict.getDocumentId();
        Utility util_obj = new Utility();
        util_obj.checkMismatchDocId(localDoc, remoteDoc, docId);
        MutableDocument newDoc = localDoc.toMutable();
        Map<String, Object> remoteDocMap = remoteDoc.toMap();
        for (Map.Entry<String, Object> entry: remoteDocMap.entrySet()) {
            String key = entry.getKey();
            Object value = entry.getValue();
            if (!newDoc.contains(key)) {
                newDoc.setValue(key, value);
            }
        }
        return newDoc;
    }
}

class IncorrectDocIdConflictResolver implements ConflictResolver {
    private static final String TAG = "CCRREPLCONFIGHANDLER";

    @Override
    public Document resolve(Conflict conflict) {
        Document localDoc = conflict.getLocalDocument();
        Document remoteDoc = conflict.getRemoteDocument();
        if (localDoc == null || remoteDoc == null) {
            throw new IllegalStateException("Either local doc or remote is/are null");
        }
        String docId = conflict.getDocumentId();
        Utility util_obj = new Utility();
        util_obj.checkMismatchDocId(localDoc, remoteDoc, docId);
        String newId = "changed" + docId;
        MutableDocument newDoc = new MutableDocument(newId, localDoc.toMap());
        newDoc.setValue("new_value", "couchbase");
        return newDoc;
    }
}

class DelayedLocalWinConflictResolver implements ConflictResolver {
    private static final String TAG = "CCRREPLCONFIGHANDLER";

    @Override
    public Document resolve(Conflict conflict) {
        Document localDoc = conflict.getLocalDocument();
        Document remoteDoc = conflict.getRemoteDocument();
        if (localDoc == null || remoteDoc == null) {
            throw new IllegalStateException("Either local doc or remote is/are null");
        }
        String docId = conflict.getDocumentId();
        Utility util_obj = new Utility();
        util_obj.checkMismatchDocId(localDoc, remoteDoc, docId);
        try { sleep(1000 * 10); } // !!!! WTF???
        catch (InterruptedException ignore) { }
        return localDoc;
    }
}

class DeleteDocConflictResolver implements ConflictResolver {
    private static final String TAG = "CCRREPLCONFIGHANDLER";

    @Override
    public Document resolve(Conflict conflict) {
        Document localDoc = conflict.getLocalDocument();
        Document remoteDoc = conflict.getRemoteDocument();
        if (localDoc == null || remoteDoc == null) {
            throw new IllegalStateException("Either local doc or remote is/are null");
        }
        String docId = conflict.getDocumentId();
        Utility util_obj = new Utility();
        util_obj.checkMismatchDocId(localDoc, remoteDoc, docId);

        return localDoc;
    }
}

class ExceptionThrownConflictResolver implements ConflictResolver {
    private static final String TAG = "CCRREPLCONFIGHANDLER";

    @Override
    public Document resolve(Conflict conflict) {
        Document localDoc = conflict.getLocalDocument();
        Document remoteDoc = conflict.getRemoteDocument();
        if (localDoc == null || remoteDoc == null) {
            throw new IllegalStateException("Either local doc or remote is/are null");
        }
        String docId = conflict.getDocumentId();
        Utility util_obj = new Utility();
        util_obj.checkMismatchDocId(localDoc, remoteDoc, docId);
        throw new IllegalStateException("Throwing an exception");
    }
}

class Utility {
    public void checkMismatchDocId(Document localDoc, Document remoteDoc, String docId) {
        String localId = localDoc.getId();
        String remoteId = remoteDoc.getId();
        if (!(localId.equals(docId) && remoteId.equals(docId))) {
            throw new IllegalStateException("Mismatch " + docId + ": local = " + localId + ", remote = " + remoteId);
        }
    }
}

File: CBLClient/Apps/JavaRequestHandlersKotlin/src/main/java/com/couchbase/mobiletestkit/javacommon/RequestHandler/ReplicatorConfigurationRequestHandler.kt
package com.couchbase.mobiletestkit.javacommon.RequestHandler

import com.couchbase.lite.*
import com.couchbase.mobiletestkit.javacommon.Args
import com.couchbase.mobiletestkit.javacommon.RequestHandlerDispatcher
import com.couchbase.mobiletestkit.javacommon.util.Log
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.InputStream
import java.net.URI
import java.net.URISyntaxException
import java.util.*

class ReplicatorConfigurationRequestHandler {
    @Throws(URISyntaxException::class)
    fun builderCreate(args: Args): ReplicatorConfiguration {
        val sourceDb = args.get<Database>("sourceDb")
        val targetDb = args.get<Database>("targetDb")
        var targetURI: URI? = null
        if (args.get<Any?>("targetURI") != null) {
            targetURI = URI(args.get<Any>("targetURI") as String)
        }
        return if (targetDb != null) {
            val target = DatabaseEndpoint(targetDb)
            ReplicatorConfiguration(sourceDb, target)
        } else if (targetURI != null) {
            val target = URLEndpoint(targetURI)
            ReplicatorConfiguration(sourceDb, target)
        } else {
            throw IllegalArgumentException("Incorrect configuration parameter provided")
        }
    }

    @Throws(Exception::class)
    fun configure(args: Args): ReplicatorConfiguration {
        val sourceDb = args.get<Database>("source_db")
        var targetURL: URI? = null
        if (args.get<Any?>("target_url") != null) {
            targetURL = URI(args.get<Any>("target_url") as String)
        }
        val targetDb = args.get<Database>("target_db")
        var replicatorType = args.get<String>("replication_type")
        val continuous = args.get<Boolean>("continuous")
        val channels = args.get<List<String>>("channels")
        val documentIds = args.get<List<String>>("documentIDs")
        val pinnedservercert = args.get<String>("pinnedservercert")
        val authenticator = args.get<Authenticator>("authenticator")
        val push_filter = args.get<Boolean>("push_filter")
        val pull_filter = args.get<Boolean>("pull_filter")
        val filter_callback_func = args.get<String>("filter_callback_func")
        val conflict_resolver = args.get<String>("conflict_resolver")
        val headers = args.get<Map<String, String>>("headers")
        val heartbeat = args.get<String>("heartbeat")
        val maxRetries = args.get<String>("max_retries")
        val maxRetryWaitTime = args.get<String>("max_timeout")
        if (replicatorType == null) {
            replicatorType = "push_pull"
        }
        replicatorType = replicatorType.toLowerCase()
        val replType: ReplicatorType
        //val replType: AbstractReplicator.ReplicatorType
        replType = if (replicatorType == "push") {
            // ReplicatorConfiguration.ReplicatorType.PUSH
            ReplicatorType.PUSH
        } else if (replicatorType == "pull") {
            // ReplicatorConfiguration.ReplicatorType.PULL
            ReplicatorType.PULL
        } else {
            // ReplicatorConfiguration.ReplicatorType.PUSH_AND_PULL
            ReplicatorType.PUSH_AND_PULL
        }
        val config: ReplicatorConfiguration
        config = if (sourceDb != null && targetURL != null) {
            val target = URLEndpoint(targetURL)
            ReplicatorConfiguration(sourceDb, target)
        } else if (sourceDb != null && targetDb != null) {
            val target = DatabaseEndpoint(targetDb)
            ReplicatorConfiguration(sourceDb, target)
        } else {
            throw Exception("\"No source db provided or target url provided\"")
        }
        if (continuous != null) {
            config.isContinuous = continuous
        } else {
            config.isContinuous = false
        }
        if (headers != null) {
            config.headers = headers
        }
        if (authenticator != null) {
            config.setAuthenticator(authenticator)
        }
        config.setType(replType)
        /*if (conflictResolver != null) {
            config.setConflictResolver(conflictResolver);
        }*/if (channels != null) {
            config.channels = channels
        }
        if (documentIds != null) {
            config.documentIDs = documentIds
        }
        if (heartbeat != null && !heartbeat.trim { it <= ' ' }.isEmpty()) {
            config.heartbeat = heartbeat.toInt()
        }
        if (maxRetries != null && !maxRetries.trim { it <= ' ' }.isEmpty()) {
            config.maxAttempts = maxRetries.toInt()
        }
        if (maxRetryWaitTime != null && !maxRetryWaitTime.trim { it <= ' ' }.isEmpty()) {
            config.maxAttemptWaitTime = maxRetryWaitTime.toInt()
        }
        Log.d(
            TAG,
            "Args: $args"
        )
        if (pinnedservercert != null) {
            val ServerCert = pinnedCertFile
            // Set pinned certificate.
            config.pinnedServerCertificate = ServerCert
        }
        if (push_filter) {
            when (filter_callback_func) {
                "boolean" -> config.pushFilter =
                    ReplicatorBooleanFilterCallback()
                "deleted" -> config.pushFilter = ReplicatorDeletedFilterCallback()
                "access_revoked" -> config.pushFilter = ReplicatorAccessRevokedFilterCallback()
                else -> config.pushFilter = DefaultReplicatorFilterCallback()
            }
        }
        if (pull_filter) {
            when (filter_callback_func) {
                "boolean" -> config.pullFilter =
                    ReplicatorBooleanFilterCallback()
                "deleted" -> config.pullFilter = ReplicatorDeletedFilterCallback()
                "access_revoked" -> config.pullFilter = ReplicatorAccessRevokedFilterCallback()
                else -> config.pullFilter = DefaultReplicatorFilterCallback()
            }
        }
        when (conflict_resolver) {
            "local_wins" -> config.conflictResolver =
                LocalWinsCustomConflictResolver()
            "remote_wins" -> config.conflictResolver = RemoteWinsCustomConflictResolver()
            "null" -> config.conflictResolver = NullCustomConflictResolver()
            "merge" -> config.conflictResolver = MergeCustomConflictResolver()
            "incorrect_doc_id" -> config.conflictResolver = IncorrectDocIdConflictResolver()
            "delayed_local_win" -> config.conflictResolver = DelayedLocalWinConflictResolver()
            "delete_not_win" -> config.conflictResolver = DeleteDocConflictResolver()
            "exception_thrown" -> config.conflictResolver = ExceptionThrownConflictResolver()
            else -> config.conflictResolver = ConflictResolver.DEFAULT
        }
        return config
    }

    fun create(args: Args): ReplicatorConfiguration {
        return args.get("configuration")
    }

    fun getAuthenticator(args: Args): Authenticator? {
        val replicatorConfiguration = args.get<ReplicatorConfiguration>("configuration")
        return replicatorConfiguration.authenticator
    }

    fun getChannels(args: Args): List<String>? {
        val replicatorConfiguration = args.get<ReplicatorConfiguration>("configuration")
        return replicatorConfiguration.channels
    }

    /*public ConflictResolver getConflictResolver(Args args){
        ReplicatorConfiguration replicatorConfiguration = args.get("configuration");
        return replicatorConfiguration.getConflictResolver();
    }*/
    fun getDatabase(args: Args): Database {
        val replicatorConfiguration = args.get<ReplicatorConfiguration>("configuration")
        return replicatorConfiguration.database
    }

    fun getDocumentIDs(args: Args): List<String>? {
        val replicatorConfiguration = args.get<ReplicatorConfiguration>("configuration")
        return replicatorConfiguration.documentIDs
    }

    fun getPinnedServerCertificate(args: Args): ByteArray? {
        val replicatorConfiguration = args.get<ReplicatorConfiguration>("configuration")
        return replicatorConfiguration.pinnedServerCertificate
    }

    fun getReplicatorType(args: Args): String {
        val replicatorConfiguration = args.get<ReplicatorConfiguration>("configuration")
        return replicatorConfiguration.replicatorType.toString()
    }

    fun getTarget(args: Args): String {
        val replicatorConfiguration = args.get<ReplicatorConfiguration>("configuration")
        return replicatorConfiguration.target.toString()
    }

    fun isContinuous(args: Args): Boolean {
        val replicatorConfiguration = args.get<ReplicatorConfiguration>("configuration")
        return replicatorConfiguration.isContinuous
    }

    fun setAuthenticator(args: Args) {
        val replicatorConfiguration = args.get<ReplicatorConfiguration>("configuration")
        val authenticator = args.get<Authenticator>("authenticator")
        replicatorConfiguration.setAuthenticator(authenticator)
    }

    fun setChannels(args: Args) {
        val replicatorConfiguration = args.get<ReplicatorConfiguration>("configuration")
        val channels = args.get<List<String>>("channels")
        replicatorConfiguration.channels = channels
    }

    /*public void setConflictResolver(Args args){
        ReplicatorConfiguration replicatorConfiguration = args.get("configuration");
        ConflictResolver conflictResolver = args.get("conflictResolver");
        replicatorConfiguration.setConflictResolver(conflictResolver);
    }*/
    fun setContinuous(args: Args) {
        val replicatorConfiguration = args.get<ReplicatorConfiguration>("configuration")
        val continuous = args.get<Boolean>("continuous")
        replicatorConfiguration.isContinuous = continuous
    }

    fun setDocumentIDs(args: Args) {
        val replicatorConfiguration = args.get<ReplicatorConfiguration>("configuration")
        val documentIds = args.get<List<String>>("documentIds")
        replicatorConfiguration.documentIDs = documentIds
    }

    fun setPinnedServerCertificate(args: Args) {
        val replicatorConfiguration = args.get<ReplicatorConfiguration>("configuration")
        val cert = args.get<ByteArray>("cert")
        replicatorConfiguration.pinnedServerCertificate = cert
    }

    fun setReplicatorType(args: Args) {
        val replicatorConfiguration = args.get<ReplicatorConfiguration>("configuration")
        val type = args.get<String>("replType")
        val replicatorType: ReplicatorType
        replicatorType = when (type) {
            /*
            "PUSH" -> ReplicatorConfiguration.ReplicatorType.PUSH
            "PULL" -> ReplicatorConfiguration.ReplicatorType.PULL
            else -> ReplicatorConfiguration.ReplicatorType.PUSH_AND_PULL*/
            "PUSH" -> ReplicatorType.PUSH
            "PULL" -> ReplicatorType.PULL
            else -> ReplicatorType.PUSH_AND_PULL
        }
        replicatorConfiguration.setType(replicatorType)
    }

    private val pinnedCertFile: ByteArray
        private get() {
            var `is`: InputStream? = null
            return try {
                `is` = RequestHandlerDispatcher.context.getAsset("sg_cert.cer")
                toByteArray(`is`)
            } finally {
                if (`is` != null) {
                    try {
                        `is`.close()
                    } catch (e: IOException) {
                    }
                }
            }
        }

    companion object {
        private const val TAG = "REPLCONFIGHANDLER"
        fun toByteArray(`is`: InputStream?): ByteArray {
            val bos = ByteArrayOutputStream()
            val b = ByteArray(1024)
            try {
                var bytesRead = `is`!!.read(b)
                while (bytesRead != -1) {
                    bos.write(b, 0, bytesRead)
                    bytesRead = `is`.read(b)
                }
            } catch (io: IOException) {
                Log.w(TAG, "Got exception " + io.message + ", Ignoring...")
            }
            return bos.toByteArray()
        }
    }
}

internal class ReplicatorBooleanFilterCallback : ReplicationFilter {
    override fun filtered(document: Document, flags: EnumSet<DocumentFlag>): Boolean {
        val key = "new_field_1"
        return if (document.contains(key)) {
            document.getBoolean(key)
        } else true
    }
}

internal class DefaultReplicatorFilterCallback : ReplicationFilter {
    override fun filtered(document: Document, flags: EnumSet<DocumentFlag>): Boolean {
        return true
    }
}

internal class ReplicatorDeletedFilterCallback : ReplicationFilter {
    override fun filtered(document: Document, flags: EnumSet<DocumentFlag>): Boolean {
        return !flags.contains(DocumentFlag.DELETED)
    }
}

internal class ReplicatorAccessRevokedFilterCallback : ReplicationFilter {
    override fun filtered(document: Document, flags: EnumSet<DocumentFlag>): Boolean {
        return !flags.contains(DocumentFlag.ACCESS_REMOVED)
    }
}

internal class LocalWinsCustomConflictResolver : ConflictResolver {
    override fun resolve(conflict: Conflict): Document? {
        val localDoc = conflict.localDocument
        val remoteDoc = conflict.remoteDocument
        check(!(localDoc == null || remoteDoc == null)) { "Either local doc or remote is/are null" }
        val docId = conflict.documentId
        val util_obj = Utility()
        util_obj.checkMismatchDocId(localDoc, remoteDoc, docId)
        return localDoc
    }

    companion object {
        private const val TAG = "CCRREPLCONFIGHANDLER"
    }
}

internal class RemoteWinsCustomConflictResolver : ConflictResolver {
    override fun resolve(conflict: Conflict): Document? {
        val localDoc = conflict.localDocument
        val remoteDoc = conflict.remoteDocument
        check(!(localDoc == null || remoteDoc == null)) { "Either local doc or remote is/are null" }
        val docId = conflict.documentId
        val util_obj = Utility()
        util_obj.checkMismatchDocId(localDoc, remoteDoc, docId)
        return remoteDoc
    }

    companion object {
        private const val TAG = "CCRREPLCONFIGHANDLER"
    }
}

internal class NullCustomConflictResolver : ConflictResolver {
    override fun resolve(conflict: Conflict): Document? {
        val localDoc = conflict.localDocument
        val remoteDoc = conflict.remoteDocument
        check(!(localDoc == null || remoteDoc == null)) { "Either local doc or remote is/are null" }
        val docId = conflict.documentId
        val util_obj = Utility()
        util_obj.checkMismatchDocId(localDoc, remoteDoc, docId)
        return null
    }

    companion object {
        private const val TAG = "CCRREPLCONFIGHANDLER"
    }
}

internal class MergeCustomConflictResolver : ConflictResolver {
    override fun resolve(conflict: Conflict): Document? {
        /**
         * Migrate the conflicted doc.
         * Algorithm creates a new doc with copying local doc and then adding any additional key
         * from remote doc. Conflicting keys will have value from local doc.
         */
        val localDoc = conflict.localDocument
        val remoteDoc = conflict.remoteDocument
        check(!(localDoc == null || remoteDoc == null)) { "Either local doc or remote is/are null" }
        val docId = conflict.documentId
        val util_obj = Utility()
        util_obj.checkMismatchDocId(localDoc, remoteDoc, docId)
        val newDoc = localDoc.toMutable()
        val remoteDocMap = remoteDoc.toMap()
        for ((key, value) in remoteDocMap) {
            if (!newDoc.contains(key)) {
                newDoc.setValue(key, value)
            }
        }
        return newDoc
    }

    companion object {
        private const val TAG = "CCRREPLCONFIGHANDLER"
    }
}

internal class IncorrectDocIdConflictResolver : ConflictResolver {
    override fun resolve(conflict: Conflict): Document? {
        val localDoc = conflict.localDocument
        val remoteDoc = conflict.remoteDocument
        check(!(localDoc == null || remoteDoc == null)) { "Either local doc or remote is/are null" }
        val docId = conflict.documentId
        val util_obj = Utility()
        util_obj.checkMismatchDocId(localDoc, remoteDoc, docId)
        val newId = "changed$docId"
        val newDoc = MutableDocument(newId, localDoc.toMap())
        newDoc.setValue("new_value", "couchbase")
        return newDoc
    }

    companion object {
        private const val TAG = "CCRREPLCONFIGHANDLER"
    }
}

internal class DelayedLocalWinConflictResolver : ConflictResolver {
    override fun resolve(conflict: Conflict): Document? {
        val localDoc = conflict.localDocument
        val remoteDoc = conflict.remoteDocument
        check(!(localDoc == null || remoteDoc == null)) { "Either local doc or remote is/are null" }
        val docId = conflict.documentId
        val util_obj = Utility()
        util_obj.checkMismatchDocId(localDoc, remoteDoc, docId)
        try {
            Thread.sleep((1000 * 10).toLong())
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
        return localDoc
    }

    companion object {
        private const val TAG = "CCRREPLCONFIGHANDLER"
    }
}

internal class DeleteDocConflictResolver : ConflictResolver {
    override fun resolve(conflict: Conflict): Document? {
        val localDoc = conflict.localDocument
        val remoteDoc = conflict.remoteDocument
        check(!(localDoc == null || remoteDoc == null)) { "Either local doc or remote is/are null" }
        val docId = conflict.documentId
        val util_obj = Utility()
        util_obj.checkMismatchDocId(localDoc, remoteDoc, docId)
        return if (remoteDoc == null) {
            localDoc
        } else {
            null
        }
    }

    companion object {
        private const val TAG = "CCRREPLCONFIGHANDLER"
    }
}

internal class ExceptionThrownConflictResolver : ConflictResolver {
    override fun resolve(conflict: Conflict): Document? {
        val localDoc = conflict.localDocument
        val remoteDoc = conflict.remoteDocument
        check(!(localDoc == null || remoteDoc == null)) { "Either local doc or remote is/are null" }
        val docId = conflict.documentId
        val util_obj = Utility()
        util_obj.checkMismatchDocId(localDoc, remoteDoc, docId)
        throw IllegalStateException("Throwing an exception")
    }

    companion object {
        private const val TAG = "CCRREPLCONFIGHANDLER"
    }
}

internal class Utility {
    fun checkMismatchDocId(localDoc: Document, remoteDoc: Document, docId: String?) {
        val remoteDocId = remoteDoc.id
        val localDocId = localDoc.id
        check(!(remoteDocId !== docId)) { "DocId mismatch" }
        check(!(docId !== localDocId)) { "DocId mismatch" }
    }
}